components/use-paid-mutation.js

```diff
@@ -32,23 +33,36 @@ export function usePaidMutation (mutation,
 
   const waitForPayment = useCallback(async (invoice, { alwaysShowQROnFailure = false, persistOnNavigate = false, waitFor }) => {
     let walletError
+    let walletInvoice = invoice
     const start = Date.now()
+
     try {
-      return await waitForWalletPayment(invoice, waitFor)
+      return await waitForWalletPayment(walletInvoice, waitFor)
     } catch (err) {
-      if (
-        (!alwaysShowQROnFailure && Date.now() - start > 1000) ||
-        err instanceof InvoiceCanceledError ||
-        err instanceof InvoiceExpiredError) {
-        // bail since qr code payment will also fail
-        // also bail if the payment took more than 1 second
-        // and cancel the invoice if it's not already canceled so it can be retried
-        invoiceHelper.cancel(invoice).catch(console.error)
+      walletError = null
+      if (err instanceof WalletError) {
+        walletError = err
+        // wallet payment error handling always creates a new invoice to retry
+        // unless no wallet was even able to attempt a payment
+        if (err.newInvoice) walletInvoice = err.newInvoice
```

Comment:
this is maybe the weirdest part of this code: using an error to pass the new invoice to the calling context

Reply:
> Can't you extract the part that recreates the invoice or return it from waitForWalletPayment ?

I don't think so. Even if I extract it, I still need to pass the latest invoice to the calling context in some way. I can't use return since that would mean the wallet payment succeeded which isn't the case here.

I do return the latest invoice on success though, see the `waitForPayment` call for pessimistic actions below.

Reply:
It might seem unimportant in the context this function is used, but these are things that build up technical debt, since now you have a function that leaks invoices, unless you remember that unexpectedly  the error contains a new resource you need to cancel if you don't need it, that's why passing around **new** data in an error field is generally bad practice, errors should report only what and why something failed.

But I think you can extract the logic and make a proper error object simply by not recreating the last invoice but just cancelling it and setting a `failedInvoice` field in the error. Then when you catch it in `waitForPayment` you can recreate it with `await invoiceHelper.retry(error.failedInvoice)`  only if it is needed.
 eg. if   `skipQr` is true, you won't even need to retry it.

Thoughts?

Reply:
Oh, I had it like this before but it looked awkward to do something different on the last payment attempt error (not create new invoice) so I changed it to this with less code.

But the argument about "new data" is a good one, I'll change it back to only save the failed invoice in the error and the caller can retry if they want so. Thanks!!

Reply:
Did this in 96f76a9.

I realized that I can filter for only the wallets that are actually able to pay so the logic if there is another wallet we can try is indeed basically just `if (i === wallets.length - 1)`. For some reason I didn't think the filter would work so I didn't even try and I think that's why I also didn't do it like this before.

This means the checks in `sendPayment` aren't really required anymore so maybe we should remove them :thinking: